### PR TITLE
Attach behavior via "as behaviorName" with Closure

### PR DESCRIPTION
**Pull request by timkelty**
Allows behaviors attached with the `'as behaviorName'` syntax to use a Closure.

In the process of writing a test, I discovered the `\yiiunit\framework\base\ComponentTest::testAttachBehavior` test was broken (it has assertions that were never reached), so I fixed that as well.

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: tests/framework/base/ComponentTest.php [hunk 1]
```diff
@@ -328,22 +328,26 @@ public function testAttachBehavior()
         $this->assertTrue($component->hasProperty('p'));
         $component->test();
         $this->assertTrue($component->behaviorCalled);
+    }
 
-        $this->assertSame($behavior, $component->detachBehavior('a'));
-        $this->assertFalse($component->hasProperty('p'));
-        $this->expectException('yii\base\UnknownMethodException');
-        $component->test();
```
Comment (timkelty) [inline tests/framework/base/ComponentTest.php:335]: The test was passing but the assertions following this were never run.

File: tests/framework/base/ComponentTest.php [hunk 2]
```diff
@@ -376,6 +380,9 @@ public function testDetachBehavior()
 
         $detachedBehavior = $component->detachBehavior('z');
         $this->assertNull($detachedBehavior);
+
+        $this->expectException('yii\base\UnknownMethodException');
+        $component->test();
```
Comment (timkelty) [inline tests/framework/base/ComponentTest.php:385]: This made more sense to test here than in `attachBehaviors` where it lived and was causing the test to end prematurely.

**Review by schmunk42 — COMMENTED** [review]
`Added lines #L193 - L194 were not covered by tests`

Could you add a test for that.

**Comment by timkelty** [comment]
> `Added lines #L193 - L194 were not covered by tests`
> 
> Could you add a test for that.

@schmunk42 the test for that is [here](https://github.com/yiisoft/yii2/pull/20248/files#diff-1329c13980fe7f57f23384af87d0f261919c4de3fca4281b4f649c19839228fdR346-R350). I'm not sure why that warning is showing in the Github diff. It isn't showing here: https://app.codecov.io/gh/yiisoft/yii2/pull/20248

File: framework/db/BaseActiveRecord.php
Comment (machour) [inline framework/db/BaseActiveRecord.php:1]: Can you rollback this change please?

**Review by machour — COMMENTED** [review]
Can you rollback changes to the two files under `db/` ?

**Comment by timkelty** [comment]
> Can you rollback changes to the two files under `db/` ?

Done. CS check will fail, but not from changes in this PR.

**Comment by mtangoo** [comment]
@timkelty thanks for your time and efforts. Can you please resolve the conflict to enable final review and possibly merging it?

**Comment by timkelty** [comment]
All set, @mtangoo.

**Comment by mtangoo** [comment]
> All set, @mtangoo.

Thanks for your time and efforts. We will review and merge it if all is good without delay!

**Review by terabytesoftw — APPROVED** [review]

**Review by bizley — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by mtangoo** [comment]
@timkelty the PR is good. Can you add a line in changelog so that we can merge this?

**Comment by timkelty** [comment]
> @timkelty the PR is good. Can you add a line in changelog so that we can merge this?

Done!

**Review by Arhell — APPROVED** [review]

**Comment by mtangoo** [comment]
@timkelty thanks for your time and efforts

**Pull request merged**